src/box/iproto.cc

```diff
@@ -2950,23 +2960,32 @@ iproto_send_listen_msg(struct evio_service *binary)
 int
 iproto_listen(const char *uri)
 {
-	if (iproto_send_stop_msg() != 0)
-		return -1;
-	evio_service_stop(&tx_binary);
-	if (uri == NULL) {
-		tx_binary.addr_len = 0;
-		return 0;
-	}
+	struct evio_service tmp;
+	memset(&tmp, 0, sizeof(tmp));
 	/*
 	 * Please note, we bind socket in main thread, and then
 	 * listen this socket in all iproto threads! With this
 	 * implementation, we rely on the Linux kernel to distribute
 	 * incoming connections across iproto threads.
 	 */
-	if (evio_service_bind(&tx_binary, uri) != 0)
+	if (uri != NULL && evio_service_bind(&tmp, uri) != 0)
 		return -1;
-	if (iproto_send_listen_msg(&tx_binary) != 0)
+	if (iproto_send_stop_msg() != 0) {
+		evio_service_stop(&tmp);
```

Comment:
If you want to test this stop, you should use existing errors, not add new. For example, add a delay to the stop (via `ERROR_INJECT_SLEEP`) branch in the iproto thread. Then `iproto_send_stop_msg` will hang. Now cancel the fiber - `cbus_call` will be terminated with an error which you will get here. This case can happen in reality.

Reply:
Fixed cbus_call, now it can't fail in case we pas TIMEOUT_INFINITY